transport_plugins/websocket/iotile_transport_websocket/virtual_websocket.py

```diff
@@ -0,0 +1,922 @@
+        """
+
+        self.send_notification(operations.NOTIFY_PROGRESS,
+                               connection_string=connection_string,
+                               done_count=done_count,
+                               total_count=total_count)
+
+    def _stream_data(self, device_uuid, chunk=None):
+        """Stream reports to the WebSocket client in 20 byte chunks
+
+        Args:
+            device_uuid (int): The uuid of the device which sent the report
+            chunk (bytes): A chunk that should be sent instead of requesting a
+                new chunk from the pending reports.
+        """
+
+        connection_string = self._uuid_to_connection_string(device_uuid)
+
+        self.streaming_data = True
+
+        if chunk is None:
+            chunk = self._next_streaming_chunk(20)
+
+        if chunk is None or len(chunk) == 0:
+            self.streaming_data = False
+            return
+
+        try:
+            self.send_notification(
+                operations.NOTIFY_REPORT,
+                connection_string=connection_string,
+                payload=base64.b64encode(chunk)
+            )
+            self._defer(self._stream_data, [device_uuid])
+        except HardwareError as err:
+            return_value = err.params['return_value']
+
+            # If we're told we ran out of memory, wait and try again
+            if return_value.get('code', 0) == 0x182:
+                time.sleep(.02)
+                self._defer(self._stream_data, [device_uuid, chunk])
+            elif return_value.get('code', 0) == 0x181:  # Invalid state, the other side likely disconnected midstream
+                self._audit('ErrorStreamingReport')
+            else:
+                self.logger.exception(err)
+                self._audit('ErrorStreamingReport')
+
+    def _send_trace(self, device_uuid, chunk=None):
+        """Stream tracing data to the WebSocket client in 20 byte chunks
+
+        Args:
+            device_uuid (int): The uuid of the device which sent the trace
+            chunk (bytes): A chunk that should be sent instead of requesting a
+                new chunk from the pending reports.
+        """
+
+        connection_string = self._uuid_to_connection_string(device_uuid)
+
+        self.tracing_data = True
+
+        if chunk is None:
+            chunk = self._next_tracing_chunk(20)
+
+        if chunk is None or len(chunk) == 0:
+            self.tracing_data = False
+            return
+
+        try:
+            self.send_notification(
+                operations.NOTIFY_TRACE,
+                connection_string=connection_string,
+                payload=base64.b64encode(chunk)
+            )
+            self._defer(self._send_trace, [device_uuid])
+        except HardwareError as err:
+            return_value = err.params['return_value']
+
+            # If we're told we ran out of memory, wait and try again
+            if return_value.get('code', 0) == 0x182:
```

Comment:
This code can be removed.  The `0x182` error code is a specific firmware response from the BLED112 dongle that it doesn't have any more free internal 20 byte buffers to queue a notification for transmission on the next ble connection event so we have to retry. 

That can never happen with websockets so this code can probably be significantly simpler since you don't have to worry about retries.